clippy_lints/src/recursive_format_impl.rs

```diff
@@ -0,0 +1,269 @@
+            then {
+                for (_i, arg) in args.iter().enumerate() {
+                    match impl_trait {
+                    // In Display, we only care about Display (it is okay to use Debug)
+                        ImplTrait::Display => {
+                    if arg.format_trait != sym::Display {
+                        continue;
+                    }},
+                    // In Debug, we only care about Debug (it is okay to use Display and ToString)
+                    ImplTrait::Debug => {
+                    if arg.format_trait != sym::Debug {
+                        continue;
+                    }},
+
+                    };
+                    check_format_arg_self(cx, expr, self_hir_id, arg, impl_trait);
+                }
+            }
+    }
+}
+
+fn check_format_arg_self(
+    cx: &LateContext<'_>,
+    expr: &Expr<'_>,
+    self_hir_id: HirId,
+    arg: &FormatArgsArg<'_>,
+    impl_trait: ImplTrait,
+) {
+    // Handle multiple dereferencing of references e.g. &&self
+    // Handle single dereference of &self -> self that is equivalent (i.e. via *self in fmt() impl)
+    // Since the argument to fmt is itself a reference: &self
+    let reference = single_deref(deref_expr(arg.value));
+    if path_to_local_id(reference, self_hir_id) {
+        match impl_trait {
+            ImplTrait::Display => {
+                span_lint(
+                    cx,
+                    RECURSIVE_FORMAT_IMPL,
+                    expr.span,
+                    "using `self` as Display in `fmt::Display` implementation might lead to infinite recursion",
+                );
+            },
+            ImplTrait::Debug => {
+                span_lint(
+                    cx,
+                    RECURSIVE_FORMAT_IMPL,
+                    expr.span,
+                    "using `self` as Debug in `fmt::Debug` implementation might lead to infinite recursion",
+                );
+            },
+        }
+    }
+}
+
+fn deref_expr<'a, 'b>(expr: &'a Expr<'b>) -> &'a Expr<'b> {
+    if let ExprKind::AddrOf(_, _, reference) = expr.kind {
+        deref_expr(reference)
+    } else {
+        expr
+    }
+}
+
+fn single_deref<'a, 'b>(expr: &'a Expr<'b>) -> &'a Expr<'b> {
+    if let ExprKind::Unary(UnOp::Deref, ex) = expr.kind {
+        ex
+    } else {
+        expr
+    }
+}
+
+fn is_format_trait_impl(cx: &LateContext<'_>, item: &Item<'_>) -> Option<ImplTrait> {
+    if_chain! {
+        // Are we at an Impl?
+        if let ItemKind::Impl(Impl { of_trait: Some(trait_ref), .. }) = &item.kind;
+        if let Some(did) = trait_ref.trait_def_id();
+        then {
+            // Is it for Display trait?
+            if match_def_path(cx, did, &paths::DISPLAY_TRAIT) {
+                Some(ImplTrait::Display)
+            }
+            // Is it for Debug trait?
+            else if match_def_path(cx, did, &paths::DEBUG_TRAIT) {
+                Some(ImplTrait::Debug)
+            } else {
+            None
```

Comment:
Prefer diagnostic items over paths whenever possible. Here you should use `cx.tcx.get_diagnostic_name`.